File: shared/web-sdk/src/connectors/inflight-message-connector.ts
import type { TurnTextDeltaMessage } from "@sentient/protocol";
import type { Connector, SentientSDKInternal } from "../connector-types.ts";
import { createLogger } from "../logger.ts";

const log = createLogger(["sentient", "sdk", "connectors", "inflight-message"]);

export interface InFlightMessage {
  readonly turnId: string;
  readonly text: string;
  /** WHICH REPLY this is. Absent against a gateway that does not stamp
   *  deltas, in which case the buffer is keyed by turn — the old behaviour. */
  readonly replyId?: string;
}

export interface InFlightMessageConnectorConfig {
  /** Called whenever ANY in-flight buffer changes (seed, delta, or clear). */
  onUpdate?: (inflight: readonly InFlightMessage[]) => void;
}

// ---------------------------------------------------------------------------
// InFlightMessageConnector — accumulates streaming assistant text for every
// turn that is still producing. Separate from committed conversation history.
//
// Capability: "message.stream"
// Direction: status (observer)
//
// Receives (2.0 wire contract):
//   - turn.started     — seed an EMPTY buffer so the UI can render its
//                        "thinking" placeholder BEFORE the first token. Without
//                        the seed the bubble stays absent through provider TFFT
//                        and the user stares at nothing.
//   - turn.text.delta  — append to that turn's buffer
//   - turn.completed   — drop that turn's buffer (committed entry follows via
//                        conversation.entry)
//   - turn.aborted     — drop that turn's buffer (the cutoff-stamped committed
//                        entry follows)
//
// MULTI-TURN (spec §7.2): the buffer is a MAP keyed by turnId, not a single
// slot. A self-initiated follow-up turn starts while the previous turn may
// still be open, and BOTH must render — two consecutive assistant bubbles with
// two turn ids is a valid, expected state. The previous single-slot design
// clobbered the still-open bubble; do not reintroduce it. Map insertion order
// IS render order.
// ---------------------------------------------------------------------------

export class InFlightMessageConnector implements Connector {
  readonly capability = "message.stream";
  readonly kind = "status" as const;

  private readonly config: InFlightMessageConnectorConfig;
  private unsubs: (() => void)[] = [];
  /** reply key → the buffer. Insertion order IS render order.
   *
   *  KEYED BY REPLY, NOT BY TURN. A ReAct turn produces text more than once
   *  and that is ONE bubble that grew — but a message the person sends mid-turn
   *  is drawn between two of those stretches, so the text after it belongs to a
   *  new bubble. The gateway decides where the boundary falls and stamps every
   *  delta (`replyId`); nothing here derives it. */
  private buffers = new Map<string, { turnId: string; text: string; replyId?: string }>();

  constructor(config: InFlightMessageConnectorConfig = {}) {
    this.config = config;
  }

  /** Every turn currently mid-stream, oldest first. Empty when idle. */
  list(): readonly InFlightMessage[] {
    return [...this.buffers.values()].map((b) => ({
      turnId: b.turnId,
      text: b.text,
      ...(b.replyId === undefined ? {} : { replyId: b.replyId }),
    }));
  }

  attach(sdk: SentientSDKInternal): void {
    this.buffers = new Map();

    this.unsubs.push(
      sdk.onMessage("turn.started", (msg: unknown) => {
        const m = msg as { turnId?: string; trigger?: string };
        if (!m.turnId || this.buffers.has(m.turnId)) return;
        // Seeded under the TURN key — `turn.started` carries no replyId, and
        // the first delta is what names the bubble. That delta re-keys this
        // placeholder in place, so it never becomes an orphan beside it.
        this.buffers.set(m.turnId, { turnId: m.turnId, text: "" });
        log.debug("turn-seeded", { turnId: m.turnId, trigger: m.trigger, inflight: this.buffers.size });
        this.emit();
      }),
    );

    this.unsubs.push(
      sdk.onMessage("turn.text.delta", (msg: unknown) => {
        const m = msg as TurnTextDeltaMessage;
        if (!m.turnId || typeof m.text !== "string") return;
        const key = m.replyId ?? m.turnId;
        // Adopt the turn-keyed placeholder ONCE, while it is still empty. A
        // non-empty one belongs to a gateway sending unstamped deltas and must
        // not be stolen.
        if (key !== m.turnId) {
          const seeded = this.buffers.get(m.turnId);
          if (seeded && seeded.text === "") {
            this.buffers.delete(m.turnId);
            this.buffers.set(key, {
              turnId: seeded.turnId,
              text: seeded.text,
              ...(m.replyId === undefined ? {} : { replyId: m.replyId }),
            });
          }
        }
        const prior = this.buffers.get(key);
        this.buffers.set(key, {
          turnId: m.turnId,
          text: (prior?.text ?? "") + m.text,
          ...(m.replyId === undefined ? {} : { replyId: m.replyId }),
        });
        this.emit();
      }),
    );

    this.unsubs.push(sdk.onMessage("turn.completed", (msg: unknown) => this.clearTurn(msg, "completed")));
    this.unsubs.push(sdk.onMessage("turn.aborted", (msg: unknown) => this.clearTurn(msg, "aborted")));
  }

  detach(): void {
    for (const unsub of this.unsubs) unsub();
    this.unsubs = [];
    this.buffers = new Map();
  }

  private clearTurn(msg: unknown, reason: "completed" | "aborted"): void {
    const m = msg as { turnId?: string };
    if (!m.turnId) return;
    // EVERY bubble of this turn: a turn the person spoke through owns more
    // than one, and clearing only the turn-keyed slot would strand the rest.
    const keys = [...this.buffers].filter(([, b]) => b.turnId === m.turnId).map(([key]) => key);
    if (keys.length === 0) return;
    for (const key of keys) this.buffers.delete(key);
    log.debug("turn-cleared", { turnId: m.turnId, reason, bubbles: keys.length, inflight: this.buffers.size });
    this.emit();
  }

  private emit(): void {
    this.config.onUpdate?.(this.list());
  }
}

File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/connectors/InFlightMessageConnector.kt
// ---------------------------------------------------------------------------
// InFlightMessageConnector — accumulates streaming assistant content while a
// turn is still producing. Exposes an in-flight slot separate from the
// committed conversation history.
//
// Mirrors web-sdk's inflight-message-connector.ts:
//   capability = "message.stream"  (status observer)
//
//   turn.started    → seed an empty buffer so the UI can render the
//                     pre-first-token "thinking" placeholder.
//   turn.text.delta → append text to THAT BUBBLE's buffer.
//   turn.completed  → commit + remove every buffer of that turn.
//   turn.aborted    → drop that turn's buffers (no commit).
//
// ONE BUFFER PER BUBBLE, NOT PER TURN. A ReAct turn produces text more than
// once — narration, a tool round trip, then the answer — and all of it belongs
// to one bubble that grew. But a message the person sends mid-turn is drawn as
// its own row BETWEEN two of those stretches, so the text after it has to start
// a new bubble. The gateway decides where that boundary falls and stamps every
// delta with a `messageId` (session-runtime.ts rotates it on a mid-turn user
// message); this connector just follows. Nothing is derived here.
//
// Threading: single-threaded; the orchestrator routes frames on one dispatcher
// and subscribes onUpdate. The mutable buffer map is owned here.
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.connectors

import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.protocol.SdkEvent
import io.sentient.mobilesdk.protocol.ServerMessage
import io.sentient.mobilesdk.sdk.ChatMessage

/** The streaming buffer for one in-flight bubble. */
data class InFlightMessage(
    val turnId: String,
    val text: String,
    /** Null against a gateway that does not stamp deltas — the buffer is then
     *  keyed by turn, which is the pre-`messageId` behaviour. */
    val messageId: String? = null,
)

class InFlightMessageConnector(
    private val onUpdate: ((InFlightMessage?) -> Unit)? = null,
    private val onEvent: ((SdkEvent) -> Unit)? = null,
) : Connector {
    override val capability: String = CAPABILITY

    private val log = createLogger("connector", "inflight-message")

    // Keyed by BUBBLE, in arrival order. A 2.0 follow-up turn (§4.5/§7.2) can open
    // while the previous turn's deltas are still landing; the pre-2.0 single `current`
    // slot silently DISCARDED the earlier turn's accumulated text. Insertion order makes
    // [inflight] the newest still-open bubble — the one the live bubble renders.
    private val buffers = LinkedHashMap<String, InFlightMessage>()

    /** Newest still-open streaming buffer; null when nothing is mid-stream. */
    fun inflight(): InFlightMessage? = buffers.values.lastOrNull()

    override fun handle(msg: ServerMessage) {
        when (msg) {
            is ServerMessage.TurnStarted -> onTurnStarted(msg.turnId)
            is ServerMessage.TurnTextDelta -> onDelta(msg.turnId, msg.messageId, msg.text)
            is ServerMessage.TurnCompleted -> onCompleted(msg.turnId)
            is ServerMessage.TurnAborted -> onAborted(msg.turnId, msg.cutoff)
            else -> Unit // not owned by this connector
        }
    }

    /** The map key for one bubble. Falls back to the turn when the gateway sent
     *  no `messageId`, which reproduces the old one-buffer-per-turn behaviour. */
    private fun keyOf(turnId: String, messageId: String?): String = messageId ?: turnId

    private fun onTurnStarted(turnId: String) {
        if (turnId.isEmpty()) return
        // Seeded under the TURN key: `turn.started` carries no messageId, and the
        // first delta is what names the bubble. `onDelta` re-keys this buffer in
        // place, so the placeholder never becomes an orphan.
        buffers[turnId] = InFlightMessage(turnId = turnId, text = "")
        log.info("seed", mapOf("turnId" to turnId, "open" to buffers.size))
        onUpdate?.invoke(inflight())
        onEvent?.invoke(SdkEvent.MessageStarted(turnId))
    }

    private fun onDelta(turnId: String, messageId: String?, text: String) {
        if (turnId.isEmpty() || text.isEmpty()) return
        val key = keyOf(turnId, messageId)

        // The seeded placeholder is keyed by turn; the first stamped delta adopts
        // it so its "thinking" bubble becomes this message's buffer rather than
        // being stranded beside it.
        if (key != turnId) adoptSeededBuffer(turnId, key, messageId)

        val isNewBubble = !buffers.containsKey(key)
        // Absent buffer → create: a delta may legitimately precede its turn.started on a
        // resume replay. Content is NEVER logged — lengths only (PrivacyGuardTest).
        val next = InFlightMessage(
            turnId = turnId,
            text = (buffers[key]?.text ?: "") + text,
            messageId = messageId,
        )
        buffers[key] = next
        if (isNewBubble && messageId != null) {
            log.info("bubble-opened", mapOf("turnId" to turnId, "messageId" to messageId, "open" to buffers.size))
            onEvent?.invoke(SdkEvent.MessageStarted(turnId, messageId))
        }
        log.debug(
            "delta",
            mapOf(
                "turnId" to turnId,
                "messageId" to (messageId ?: "-"),
                "deltaLen" to text.length,
                "totalLen" to next.text.length,
                "open" to buffers.size,
            ),
        )
        onUpdate?.invoke(inflight())
        // One event per chunk — never batched.
        onEvent?.invoke(SdkEvent.MessageDelta(turnId = turnId, chunk = text, messageId = messageId))
    }

    /** Move the turn-keyed placeholder onto its real bubble key, ONCE — only
     *  while it is still empty. A non-empty turn-keyed buffer belongs to a
     *  gateway that sent unstamped deltas and must not be stolen. */
    private fun adoptSeededBuffer(turnId: String, key: String, messageId: String?) {
        val seeded = buffers[turnId] ?: return
        if (seeded.text.isNotEmpty()) return
        buffers.remove(turnId)
        buffers[key] = seeded.copy(messageId = messageId)
    }

    private fun onCompleted(turnId: String) {
        // EVERY bubble of this turn, not just the newest: a turn the person spoke
        // through owns more than one, and each is a reply that must be committed.
        val done = buffers.entries.filter { it.value.turnId == turnId }.map { it.key to it.value }
        if (done.isEmpty()) return // unknown turn → no-op; committing "" would drain a peer's live bubble
        for ((key, _) in done) buffers.remove(key)
        log.info("done", mapOf("turnId" to turnId, "bubbles" to done.size, "open" to buffers.size))
        onUpdate?.invoke(inflight())
        for ((_, buffer) in done) {
            // ts=0: clear-signal; committed text is authoritative via feed/timeline.
            onEvent?.invoke(
                SdkEvent.MessageCommitted(
                    ChatMessage(
                        ts = 0,
                        role = "assistant",
                        content = buffer.text,
                        streaming = false,
                        turnId = turnId,
                        messageId = buffer.messageId,
                    ),
                ),
            )
        }
    }

    private fun onAborted(turnId: String, cutoff: String) {
        val dropped = buffers.entries.filter { it.value.turnId == turnId }.map { it.key }
        if (dropped.isEmpty()) return
        for (key in dropped) buffers.remove(key)
        log.info("aborted", mapOf("turnId" to turnId, "cutoff" to cutoff, "dropped" to dropped.size))
        onUpdate?.invoke(inflight())
        // No MessageCommitted on abort; the abort surfaces as SdkEvent.TurnAborted (TurnErrorConnector).
    }

    companion object {
        const val CAPABILITY: String = "message.stream"
    }
}
